exonum/tests/collections/main.rs

```diff
@@ -14,111 +14,50 @@
 
 //! Tests that compare exonum collections and corresponding rust types using proptest.
 
-extern crate exonum;
 #[macro_use]
 extern crate proptest;
+extern crate exonum;
 extern crate modifier;
 
-use modifier::Modifier;
-use std::collections::HashMap;
-
-mod list_index;
-mod map_index;
-mod proof_list_index;
-mod proof_map_index;
-
-// Max size of the generated sequence of actions.
-const ACTIONS_MAX_LEN: usize = 100;
-
-#[derive(Debug, Clone)]
-enum ListAction<V> {
-    Push(V),
-    Pop,
-    Extend(Vec<V>),
-    // Applied with argument modulo `collection.len()`.
-    Truncate(u64),
-    // Applied to index modulo `collection.len()`.
-    Set(u64, V),
-    Clear,
-    MergeFork,
-}
-
-#[derive(Debug, Clone)]
-enum MapAction<K, V> {
-    // Should be applied to a small subset of keys (like modulo 8 for int).
-    Put(K, V),
-    // Should be applied to a small subset of keys (like modulo 8 for int).
-    Remove(K),
-    Clear,
-    MergeFork,
-}
-
-impl<V> Modifier<Vec<V>> for ListAction<V> {
-    fn modify(self, list: &mut Vec<V>) {
-        match self {
-            ListAction::Push(val) => {
-                list.push(val);
-            }
-            ListAction::Pop => {
-                list.pop();
-            }
-            ListAction::Extend(vec) => {
-                list.extend(vec);
-            }
-            ListAction::Truncate(size) => {
-                let len = list.len();
-                if len > 0 {
-                    list.truncate(size as usize % len);
-                }
-            }
-            ListAction::Set(idx, val) => {
-                let len = list.len();
-                if len > 0 {
-                    list[idx as usize % len] = val;
+macro_rules! proptest_compare_collections {
```

Comment:
It's a bit unhygienic to expect the macro expansion environment to contain `compare_collections()` and `generate_action()` functions with correct signatures.

And it seems that this macro could be replaced with a function generic over an impl of

```rust
trait CollectionComparison {
    type Index;
    type Reference;

    fn new_index(&str, &mut Fork) -> Self::Index;
    fn new_reference() -> Self::Reference;
    fn generate_actions() -> Box<dyn Strategy<Value = ListAction<i32>>>;
    fn compare_collections(&Self::Index, &Self::Reference) -> TestCaseResult;
};
```

or something. However, that's my personal preference of avoiding macros for non-syntax operations.

Reply:
Thanks, and I agree.
But it looks like your proposal will be verbose to implement because there are two things that conflict.
1) If you use associated types for Index it is nice since you don't have to specify traits for types but then you can't satisfy lifetimes for index(afaik there are no HRTB for associated types see this https://github.com/rust-lang/rfcs/blob/master/text/1598-generic_associated_types.md)
2) If you use generics to solve lifetime issue with HRTB then you have to specify traits for Index type. It may be hard to do this since there are too many things that goes in different ways for different collections => it may be unreadable.

Reply:
Yeah... If using generics is more verbose and less readable then macros then I don't see any issue with using macros for the job. Lifetimes and type inference are hard.

Thank you for the explanation.